### PR TITLE
fix: Cve 2017 5982

### PR DESCRIPTION
**Pull request by wsnipex**
## Description

Define white and blacklists for files/dirs allowed in webserver.
Still WIP, please review and test.

## Motivation and Context

https://forum.kodi.tv/showthread.php?tid=335695&pid=2776984#pid2776984

this is based on work by MrMC. Windows is not implemented yet, can someone please help?

## How Has This Been Tested?

Pretty sure this will break some chorus functionality.
I see requests that need to be fixed, e.g.:
ERROR: CheckFileAccessAllowed denied access to image://https://api-tvthek.orf.at/uploads/media/channels/0001/01/thumb_11_channels_list.png/

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [x] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/utils/FileUtils.cpp [hunk 1]
```diff
@@ -201,3 +203,36 @@ CDateTime CFileUtils::GetModificationDate(const std::string& strFileNameAndPath,
...
+#if defined(TARGET_POSIX)
+  char *fullpath = realpath(filePath.c_str(), nullptr);
+  if (fullpath)
+  {
+    const std::string testpath = fullpath;
+    free(fullpath);
+    CLog::Log(LOGDEBUG,"%s filePath: %s decodePath: %s realpath: %s",  __FUNCTION__, filePath.c_str(), decodePath.c_str(), testpath.c_str());
+    // allow kodi paths
+    // TODO: most likely many more needed
+    std::string userHome = CSpecialProtocol::TranslatePath("special://home");
```
Comment (notspiff) [inline xbmc/utils/FileUtils.cpp]: make this a string list for easier extension.

Comment (wsnipex) [reply xbmc/utils/FileUtils.cpp]: did you check the 2nd commit?

Comment (notspiff) [reply xbmc/utils/FileUtils.cpp]: aparently not!

File: xbmc/utils/FileUtils.cpp [hunk 2]
```diff
@@ -201,3 +203,59 @@ CDateTime CFileUtils::GetModificationDate(const std::string& strFileNameAndPath,
   }
   return dateAdded;
 }
+
+bool CFileUtils::CheckFileAccessAllowed(const std::string &filePath)
```
Comment (DaveTBlake) [inline xbmc/utils/FileUtils.cpp]: Don't we already have  `CFileUtils::RemoteAccessAllowed` to do this?

File: xbmc/utils/FileUtils.cpp [hunk 3]
```diff
@@ -201,3 +203,59 @@ CDateTime CFileUtils::GetModificationDate(const std::string& strFileNameAndPath,
...
+{
+  // DENY access to paths matching
+  const std::vector<std::string> blacklist = {
+    "passwords.xml",
+    "sources.xml",
+    "guisettings.xml",
+    "advancedsettings.xml",
+    ".ssh",
+    "/etc"
```
Comment (DaveTBlake) [inline xbmc/utils/FileUtils.cpp]: Surely "/etc" could be found in valid folder paths too

File: xbmc/utils/FileUtils.cpp (hunk 2 above)
Comment (wsnipex) [reply xbmc/utils/FileUtils.cpp]: RemoteAccessAllowed doesn't properly validate the real path. I opted not to change that for v18, but add an additional check instead

File: xbmc/utils/FileUtils.cpp (hunk 3 above)
Comment (wsnipex) [reply xbmc/utils/FileUtils.cpp]: debateable. If you can show an example, I'll remove it

Comment (notspiff) [reply xbmc/utils/FileUtils.cpp]: remember this is a find, not a startswith wsnipex

/home/etcetera/

Comment (wsnipex) [reply xbmc/utils/FileUtils.cpp]: /etc/ better?

Comment (DaveTBlake) [reply xbmc/utils/FileUtils.cpp]: Yeap, and "etc" itself is the name of several artists so could easily be a folder name

File: xbmc/utils/FileUtils.cpp (hunk 2 above)
Comment (DaveTBlake) [reply xbmc/utils/FileUtils.cpp]: I don't see what is wrong with validation done by `RemoteAccessAllowed`, doesn't it just need calling earlier plus add the black list if we want one. If the additional check blocks more than `RemoteAccessAllowed` then it makes RemoteAccessAllowed redundant

File: xbmc/utils/FileUtils.cpp (hunk 3 above)
Comment (pkerling) [reply xbmc/utils/FileUtils.cpp]: Why do we have to put `/etc` separately on the blacklist? Isn't it supposed to be blocked anyway due to not being under any of the whitelisted locations?

File: xbmc/utils/FileUtils.cpp [hunk 4]
```diff
@@ -201,3 +203,59 @@ CDateTime CFileUtils::GetModificationDate(const std::string& strFileNameAndPath,
...
+  for (const auto &b : blacklist)
+  {
+    if (decodePath.find(b) != std::string::npos)
+    {
+      CLog::Log(LOGERROR,"%s denied access to %s",  __FUNCTION__, decodePath.c_str());
+      return false;
+    }
+  }
+
+#if defined(TARGET_POSIX)
```
Comment (pkerling) [inline xbmc/utils/FileUtils.cpp]: No whitelist check for non-POSIX?

Comment (wsnipex) [reply xbmc/utils/FileUtils.cpp]: let me quote the PR description: "Windows is not implemented yet, can someone please help?"

Comment (pkerling) [reply xbmc/utils/FileUtils.cpp]: Reading is an underappreciated skill. Next time just hit me please :bowing_man: 

That said, wouldn't it be enough to replace the `realpath` part on other platforms? The logic itself should stay the same, right?

**Comment by pkerling** [comment]
> I see requests that need to be fixed, e.g.:
ERROR: CheckFileAccessAllowed denied access to image://api-tvthek.orf.at/uploads/media/channels/0001/01/thumb_11_channels_list.png

So we will probably have to allow http(s) to get any covert art.
There is another vulnerability lurking in that this allows to use the host as a proxy for arbitrary GET requests, but I don't think we will be able to fix that for v18. You'd need a different URL scheme that accesses covert art by database id (or sth similar) and not by http URIs.

**Comment by wsnipex** [comment]
indeed. stuff like image://http://some.random.host/ gives me the creeps from a security pov.

**Comment by wsnipex** [comment]
updated. We now check image:// stuff only if we can resolve it to a locally existing file, otherwise we let it through.

**Comment by wsnipex** [comment]
jenkins build this please

@Rechi could you implement the needed changes for windows?

File: xbmc/utils/FileUtils.cpp [hunk 5]
```diff
@@ -5,6 +5,8 @@
  *  SPDX-License-Identifier: GPL-2.0-or-later
  *  See LICENSES/README.md for more information.
  */
+#include <vector>
```
Comment (pkerling) [inline xbmc/utils/FileUtils.cpp]: system includes should come after kodi headers

File: xbmc/utils/FileUtils.cpp [hunk 6]
```diff
@@ -201,3 +203,77 @@ CDateTime CFileUtils::GetModificationDate(const std::string& strFileNameAndPath,
...
+    "passwords.xml",
+    "sources.xml",
+    "guisettings.xml",
+    "advancedsettings.xml",
+    "/.ssh/",
+  };
+  // ALLOW kodi paths
+  // TODO: most likely many more needed
+  const std::vector<std::string> whitelist = {
+    CSpecialProtocol::TranslatePath("special://home"),
```
[the remark this answers is not in the document]
Comment (wsnipex) [reply xbmc/utils/FileUtils.cpp]: you'll break tons of addons without that.

[the remark this answers is not in the document]
Comment (pkerling) [reply xbmc/utils/FileUtils.cpp]: Well what do they access, concretely?

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by wsnipex** [comment]
did anyone test this?

**Comment by MilhouseVH** [comment]
This has been in my RPi/Generic LibreELEC test builds since 2 Oct, so a few thousand daily users, and no complaints so far.

I did test vulnerable paths (`/etc/passwd`, `/storage/.kodi/userdata/passwords.xml` etc.) and could no longer exploit Kodi on Linux with this PR applied.

:+1:

**Comment by wsnipex** [comment]
I would like to get this in until the weekend. Anything important missing?

**Comment by pkerling** [comment]
I'll test today or tomorrow and see if I can find something missing, but looks good so far. We can implement stricter checks after v18.

Did you (or anyone really) already test whether the option of making library sources available still works?

Also @Rechi did you test accessing `passwords.xml` and system/home dir stuff on Windows?

**Comment by Rechi** [comment]
`passwords.xml` is already handled by the generic (non platform specific) code. Tested to access folders not added as source and it worked (not seeing the content).

**Pull request merged**